### DISCUSSION TITLE
Issue with Transforming Transpose and Mul Layers in CNN Model

### DISCUSSION THREAD

**Discussion by Nishith0402** [Xilinx/finn#1200] [General]
I am working with a CNN model in ONNX format. However, I am facing an issue with the transformation of a Transpose layer in my CNN model in ONNX format. After applying the streamline and convert_to_hw transformations in FINN, a Transpose layer is automatically added to convert the data format from channels-first (NCHW) to channels-last (NHWC). Even when I modify the ONNX model to use the channels-last (NHWC) format from the beginning, the Transpose layer still gets added during the transformation steps.

Additionally,there is another Transpose layer followed by a Flatten layer, and I was unable to absorb the Transpose layer into the Flatten layer using "_**AbsorbTransposeIntoFlatten()**_ ".And also I am  encountering an issue at the final stage where a Mul layer needs to be converted into a Thresholding layer, but the transformation does not take place as expected.After the convert_to_hw transformation, the create_dataflow_partition function throws an error: _**"cycle-free graph violated: partition depends on itself."**_

**Initial Transpose Layer:**
![Screenshot from 2024-09-26 18-43-38](https://github.com/user-attachments/assets/1660c211-c273-4404-8116-ca21d330bb14)

**Transpose layer followed by flatten:** 
![Screenshot from 2024-09-26 18-44-06](https://github.com/user-attachments/assets/544ebb37-559a-44e4-855c-68a9f79f0f32)

Is there a way to prevent streamline and convert_to_hw from automatically adding this Transpose layer and convert mul layer to thresholding layer at the output?

I’m looking for suggestions on how to manage these transformations and optimize the affected layers more efficiently using FINN.

**Comment by fpjentzsch** [comment]
Hi,

1) FINN cannot implement the initial Transpose and the input quantization MultiThreshold (I assume it has a floating point input?), so we usually perform them in software before passing the data to the FINN accelerator. You can either remove them manually or leave them in the graph if FINN cannot map them to hw_custom_ops, because FINN will silently cut off unsupported operations at the beginning and end of the graph, while failing with "cycle-free graph violated: partition depends on itself" if unsupported operations sit in between supported operations.
2) For this convolution -> flatten -> dense layer scenario you should try the [RemoveCNVtoFCFlatten()](https://github.com/Xilinx/finn/blob/9d299689f2ec0895f208b8bfe3bcdcf6f450181a/src/finn/transformation/move_reshape.py#L9) transformation. In your case it will probably fail because it requires an MVAU after the Flatten, not a Thresholding. I believe this Thresholding in your graph is redundant, since you already have a Thresholding before the MaxPool. Your first screenshot also shows a series of two redundant Thresholding layers.
3) There is no MultiThreshold layer at the output into which the Mul could be merged. Since it is floating point, you'll need to perform this operation outside of FINN in software or change your model to remove it or add an additional quantizer to the output if possible.

**Reply by Nishith0402** [reply]
I used quantization aware training using brevitas with input_quant=Int8ActPerTensorFloat in QuantConv2d. So, I assume even giving the input as floating point it will automatically quantize the inputs which I observed when done the verification of onnx model.

The transpose layers are inserted once the step_streamline transformation is applied. The LowCovtoMatMul transformation introduced those transpose and flatten nodes.I checked it's implementation and observed that it converts the pytorch model which is basically the channels first to channels last. And I assume that's the reason for transpose and flatten nodes insertion in th onnx graph. I am not sure why redundant thresholding node is inserted.

For the case of mul node i removed it from the onnx graph node i assume it will work.

**Comment by KanishR1** [comment]
@fpjentzsch I am @Nishith0402 's teammate in the project. As you suggested, I have modified my CNN. Now for estimating the reports, all the transformations are working. But I observed that the input and output layers are removed when the onnx transformation happens as mentioned by you earlier.

After tidy up transformation my onnx has an input layer
![Screenshot from 2024-09-27 16-16-55](https://github.com/user-attachments/assets/24f8655e-7c50-40f8-8039-4951279e237e)


After streamline transformation transpose is introduced in the input. Could you explain why this is happening and suggest any modifications to prevent this.

![Screenshot from 2024-09-27 15-21-16](https://github.com/user-attachments/assets/ae41b29d-a987-48e1-92e3-9bcf8a7788d8)

![Screenshot from 2024-09-27 16-17-07](https://github.com/user-attachments/assets/42a45567-1d6a-4177-80a1-802299375a26)

Also after creating the dataflow partition my input and output layer are removed as shown below: (I suppose it is because of the transpose in the input and add matmul and mul in the output which cause this removal. Any suggestions to prevent this)

![Screenshot from 2024-09-27 16-16-39](https://github.com/user-attachments/assets/8b063cce-7b7d-4457-ab73-6a1ec9a2639a)

![Screenshot from 2024-09-27 16-16-34](https://github.com/user-attachments/assets/404a9ccf-cc38-4876-bb43-1a4606c56938)

So, I don't understand why my input and output layers are removed. Please help us out in understanding why these are happening.

Also, when I try to generate hardware, I get an hls error

**Reply by fpjentzsch** [reply]
FINN removes all layers at the beginning/end of the graph which it cannot map to a hw backend layer, such as floating point operations or operations with unsupported hyperparameters.

As for the initial Transpose. Normally you want the Transpose here, because you train in PyTorch with the default NCHW data layout but FINN uses NHWC internally. You simply leave the Transpose, it doesn't get implemented by FINN, and you transpose the data yourself before you feed it to the accelerator.

Now if you actually trained/exported in NHWC already, you might need to manually annotate the data layout of the input tensor. By default, FINN will always assume the input is NCHW and you will see the following warning in your build_dataflow.log:

> [2022-07-21 09:47:18,588] /home/felixj/WD/finn/deps/qonnx/src/qonnx/transformation/infer_data_layouts.py:119: UserWarning: Assuming 4D input is NCHW
[2022-07-21 09:47:18,588]   warnings.warn("Assuming 4D input is NCHW")

**Reply by KanishR1** [reply]
Finn removes transpose as well as the input conv2d layer. How do I prevent this ?
